cli/flox-rust-sdk/src/models/lockfile.rs

```diff
@@ -1,37 +1,198 @@
+            pkgdb_cmd.arg("--lockfile").arg(canonical_lockfile_path);
+        }
+
+        debug!("locking manifest with command: {pkgdb_cmd:?}");
+        call_pkgdb(pkgdb_cmd)
+            .map_err(LockedManifestError::LockManifest)
+            .map(Self)
+    }
+
+    /// Build a locked manifest
+    ///
+    /// if a gcroot_out_link_path is provided,
+    /// the environment will be linked to that path and a gcroot will be created
+    pub fn build(
+        &self,
+        pkgdb: &Path,
+        gcroot_out_link_path: Option<&Path>,
+    ) -> Result<PathBuf, LockedManifestError> {
+        let mut pkgdb_cmd = Command::new(pkgdb);
+        pkgdb_cmd.arg("buildenv").arg(&self.to_string());
+
+        if let Some(gcroot_out_link_path) = gcroot_out_link_path {
+            pkgdb_cmd.args(["--out-link", &gcroot_out_link_path.to_string_lossy()]);
+        }
+
+        debug!("building environment with command: {pkgdb_cmd:?}");
+
+        let result: BuildEnvResult =
+            serde_json::from_value(call_pkgdb(pkgdb_cmd).map_err(LockedManifestError::BuildEnv)?)
+                .map_err(LockedManifestError::ParseBuildEnvOutput)?;
+
+        Ok(PathBuf::from(result.store_path))
+    }
+}
+
+impl ToString for LockedManifest {
+    fn to_string(&self) -> String {
+        self.0.to_string()
+    }
+}
+
 /// An environment (or global) lockfile.
 ///
-/// The authoritative definition of this structure is in C++.
+/// This struct is meant **for reading only**.
+///
+/// It serves as a typed representation of the lockfile json produced by pkgdb.
+/// Parsing of the lockfile is done in [TypedLockedManifest::try_from]
+/// and should be as late as possible.
+/// Where possible, use the opaque [LockedManifest] instead of this struct
+/// to avoid incompatibility issues with the authoritative definition in C++.
+///
+/// In the optimal case the lockfile schema can be inferred from a common
+/// or `pkgdb`-defined schema.
+///
 /// This struct is used as the format to communicate with pkgdb.
 /// Many pkgdb commands will need to pass some of the information in the
 /// lockfile through to Rust.
+///
+/// And some commands (i.e. `list`) will need to read lockfiles
+/// to get information about the environment without having to call `pkgdb`.
+///
 /// Although we could selectively pass fields through,
 /// I'm hoping it will be easier to parse the entirety of the lockfile in Rust,
 /// rather than defining a separate set of fields for each different pkgdb
 /// command.
-#[derive(Deserialize, Serialize)]
-pub struct Lockfile {
-    pub registry: Registry,
-    #[serde(flatten)]
-    _json: Value,
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+pub struct TypedLockedManifest {
+    #[serde(rename = "lockfile-version")]
+    lockfile_version: Version<0>,
+    packages: BTreeMap<System, BTreeMap<String, LockedPackage>>,
+    registry: Registry,
+}
+
+#[derive(Debug, Serialize, Clone, Deserialize, PartialEq)]
```

Comment:
suggestion nonblocking: drop `Serialize` on all of these since they're meant to be read only